modules/apps/site/site-cms-site-initializer/src/main/java/com/liferay/site/cms/site/initializer/internal/security/permission/resource/ObjectDefinitionModelResourcePermissionTracker.java

```diff
@@ -0,0 +1,167 @@
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+/**
+ * @author Jürgen Kappler
+ */
+@Component(service = {})
+public class ObjectDefinitionModelResourcePermissionTracker {
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTracker = ServiceTrackerFactory.open(
+			bundleContext,
+			(Class<ModelResourcePermission<?>>)
+				(Class<?>)ModelResourcePermission.class,
+			new ModelResourcePermissionServiceTrackerCustomizer(bundleContext));
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		_serviceTracker.close();
+	}
+
+	private static final Log _log = LogFactoryUtil.getLog(
+		ObjectDefinitionModelResourcePermissionTracker.class);
+
+	@Reference
+	private ObjectDefinitionLocalService _objectDefinitionLocalService;
+
+	@Reference
+	private ResourceActionLocalService _resourceActionLocalService;
+
+	@Reference
+	private ResourcePermissionLocalService _resourcePermissionLocalService;
+
+	private ServiceTracker
+		<ModelResourcePermission<?>, ModelResourcePermission<?>>
+			_serviceTracker;
+
+	private class ModelResourcePermissionServiceTrackerCustomizer
+		implements ServiceTrackerCustomizer
+			<ModelResourcePermission<?>, ModelResourcePermission<?>> {
+
+		public ModelResourcePermissionServiceTrackerCustomizer(
+			BundleContext bundleContext) {
+
+			_bundleContext = bundleContext;
+		}
+
+		@Override
+		public ModelResourcePermission<?> addingService(
+			ServiceReference<ModelResourcePermission<?>> serviceReference) {
+
+			_addResourcePermissions(serviceReference);
+
+			return _bundleContext.getService(serviceReference);
+		}
+
+		@Override
+		public void modifiedService(
+			ServiceReference<ModelResourcePermission<?>> serviceReference,
+			ModelResourcePermission<?> modelResourcePermission) {
+		}
+
+		@Override
+		public void removedService(
+			ServiceReference<ModelResourcePermission<?>> serviceReference,
+			ModelResourcePermission<?> infoItemCapabilitiesProvider) {
+
+			_bundleContext.ungetService(serviceReference);
+		}
+
+		private void _addResourcePermissions(
+			ServiceReference<ModelResourcePermission<?>> serviceReference) {
+
+			if (!GetterUtil.getBoolean(
+					serviceReference.getProperty("com.liferay.object"))) {
+
+				return;
+			}
+
+			long companyId = GetterUtil.getLong(
+				serviceReference.getProperty("companyId"));
+
+			if (!FeatureFlagManagerUtil.isEnabled(companyId, "LPD-17564")) {
+				return;
+			}
+
+			String className = GetterUtil.getString(
+				serviceReference.getProperty("model.class.name"));
+
+			try {
+				ObjectDefinition objectDefinition =
+					_objectDefinitionLocalService.
+						fetchObjectDefinitionByClassName(companyId, className);
+
+				if (objectDefinition == null) {
+					return;
+				}
+
+				Role role = CMSRoleUtil.addAdminRoleAndPermissions(companyId);
```

Comment:
Same here: why not create it when creating the company? (I'm sure I'm missing something ...).